Could not install 27.2 on Apple Silicon M1

## What you were trying to do

Previously installed 27.1 fine on this M1 machine, but today upgrading to 27.2 on the same machine failed. Uninstalled `emacs-plus@27` first, then tried installing `emacs-plus@27` again.

## What happened (include command output)

<pre>
$ arch -arm64 brew uninstall emacs-plus@27
$ arch -arm64 brew install emacs-plus@27 --with-emacs-card-british-racing-green-icon
</pre>

<details>
  <summary>Command output</summary>
  <pre>

  ==> Installing emacs-plus@27 from d12frosted/emacs-plus
==> Downloading https://raw.githubusercontent.com/d12frosted/homebrew-emacs-plus/master/patches/emacs-27/fix-window-role.patch
Already downloaded: /Users/cyan/Library/Caches/Homebrew/downloads/3d1bf46a0adb1be574f7bc46e66e7f420cb67ba348bcfda4ded5c1b61ec3ee84--fix-window-role.patch
==> Downloading https://raw.githubusercontent.com/d12frosted/homebrew-emacs-plus/master/patches/emacs-27/system-appearance.patch
Already downloaded: /Users/cyan/Library/Caches/Homebrew/downloads/84c196d082a34b90b086e387fd035a09ac07c4e2a8794dd1dd5203ccec42e158--system-appearance.patch
==> Downloading https://raw.githubusercontent.com/d12frosted/homebrew-emacs-plus/master/patches/emacs-27/ligatures-freeze-fix.patch
Already downloaded: /Users/cyan/Library/Caches/Homebrew/downloads/563e8387190e9139d2d182ba789cf1a64338a14cb300c3584162eb5482c3ff7e--ligatures-freeze-fix.patch
==> Downloading https://raw.githubusercontent.com/d12frosted/homebrew-emacs-plus/master/icons/emacs-card-british-racing-green.icns
Already downloaded: /Users/cyan/Library/Caches/Homebrew/downloads/f2b8c760dc98d9d177434e6648c621a198d9aaf0337ed03f7e4a2d9b160294bf--emacs-card-british-racing-green.icns
==> Downloading https://ftp.gnu.org/gnu/emacs/emacs-27.2.tar.xz
Already downloaded: /Users/cyan/Library/Caches/Homebrew/downloads/2cd8fd48748257712575a817f0cb133837fd4fbd6f5d6f7c402401af9f862e32--emacs-27.2.tar.xz
==> Patching
==> Applying fix-window-role.patch
patching file src/nsterm.m
Hunk #1 succeeded at 8521 (offset -247 lines).
==> Applying system-appearance.patch
patching file src/frame.h
patching file src/nsfns.m
patching file src/nsterm.m
Hunk #1 succeeded at 2029 (offset 2 lines).
Hunk #2 succeeded at 5587 (offset 2 lines).
Hunk #3 succeeded at 5833 (offset 2 lines).
Hunk #4 succeeded at 5883 (offset 2 lines).
Hunk #5 succeeded at 6110 (offset 2 lines).
Hunk #6 succeeded at 7594 (offset -1 lines).
Hunk #7 succeeded at 9723 (offset -1 lines).
==> Applying ligatures-freeze-fix.patch
patching file src/composite.c
Hunk #1 succeeded at 1769 (offset 23 lines).
==> ImageMagick PKG_CONFIG_PATH:
/opt/homebrew/opt/imagemagick/lib/pkgconfig
==> ./autogen.sh
==> ./configure --enable-locallisppath=/opt/homebrew/share/emacs/site-lisp --infodir=/opt/homebrew/Cellar/emacs-plus@27/27.2/share/info/emacs --prefix=/opt/homebrew/Cellar/emacs-plus@27/27.2 --with-xml2 -
==> make
Last 15 lines from /Users/cyan/Library/Logs/Homebrew/emacs-plus@27/03.make:
clang -o temacs.tmp \
          -Demacs  -I. -I. -I../lib -I../lib      -I/opt/homebrew/Cellar/libffi/3.3_3/include -I/opt/homebrew/Cellar/libpng/1.6.37/include/libpng16 -I/opt/homebrew/opt/freetype/include/freetype2 -I/opt/homebrew/Cellar/fontconfig/2.13.1/include -I/opt/homebrew/Cellar/pcre/8.44/include -I/opt/homebrew/Cellar/glib/2.68.0/include -I/opt/homebrew/Cellar/glib/2.68.0/include/glib-2.0 -I/opt/homebrew/Cellar/glib/2.68.0/lib/glib-2.0/include -I/opt/homebrew/opt/gettext/include -I/opt/homebrew/Cellar/libxau/1.0.9/include -I/opt/homebrew/Cellar/libxdmcp/1.1.3/include -I/opt/homebrew/Cellar/libxcb/1.14_1/include -I/opt/homebrew/Cellar/libx11/1.7.0/include -I/opt/homebrew/Cellar/libxext/1.3.4/include -I/opt/homebrew/Cellar/libxrender/0.9.10/include -I/opt/homebrew/Cellar/pixman/0.40.0/include/pixman-1 -I/opt/homebrew/Cellar/cairo/1.16.0_5/include/cairo -I/opt/homebrew/Cellar/gdk-pixbuf/2.42.4/include/gdk-pixbuf-2.0 -I/opt/homebrew/Cellar/librsvg/2.50.3/include/librsvg-2.0 -I/opt/homebrew/Cellar/xorgproto/2021.3/include -Xpreprocessor -fopenmp -DMAGICKCORE_HDRI_ENABLE=1 -DMAGICKCORE_QUANTUM_DEPTH=16 -Xpreprocessor -fopenmp -DMAGICKCORE_HDRI_ENABLE=1 -DMAGICKCORE_QUANTUM_DEPTH=16 -I/opt/homebrew/Cellar/imagemagick/7.0.11-5/include/ImageMagick-7         -I/opt/homebrew/Cellar/little-cms2/2.12/include       -MMD -MF deps/.d -MP  -I/opt/homebrew/Cellar/jansson/2.13.1/include -I/opt/homebrew/Cellar/libidn2/2.3.0/include -I/opt/homebrew/Cellar/libtasn1/4.16.0_1/include -I/opt/homebrew/Cellar/nettle/3.7.2/include -I/opt/homebrew/Cellar/p11-kit/0.23.22/include/p11-kit-1 -I/opt/homebrew/Cellar/gnutls/3.6.15/include    -Wno-switch -Wno-pointer-sign -Wno-string-plus-int -Wno-unknown-attributes -Wno-initializer-overrides -Wno-tautological-compare -Wno-tautological-constant-out-of-range-compare -g3 -O2  -framework AppKit -framework IOKit -framework Carbon  \
            dispnew.o frame.o scroll.o xdisp.o menu.o  window.o charset.o coding.o category.o ccl.o character.o chartab.o bidi.o cm.o term.o terminal.o xfaces.o    emacs.o keyboard.o macros.o keymap.o sysdep.o bignum.o buffer.o filelock.o insdel.o marker.o minibuf.o fileio.o dired.o cmds.o casetab.o casefiddle.o indent.o search.o regex-emacs.o undo.o alloc.o pdumper.o data.o doc.o editfns.o callint.o eval.o floatfns.o fns.o font.o print.o lread.o dynlib.o emacs-module.o syntax.o  bytecode.o process.o gnutls.o callproc.o region-cache.o sound.o timefns.o atimer.o doprnt.o intervals.o textprop.o composite.o xml.o lcms.o kqueue.o  profiler.o decompress.o thread.o systhread.o        fontset.o fringe.o image.o  json.o  nsterm.o nsfns.o nsmenu.o nsselect.o nsimage.o macfont.o terminfo.o lastfile.o      ../lib/libgnu.a          -L/opt/homebrew/Cellar/glib/2.68.0/lib -L/opt/homebrew/opt/gettext/lib -L/opt/homebrew/Cellar/cairo/1.16.0_5/lib -L/opt/homebrew/Cellar/gdk-pixbuf/2.42.4/lib -L/opt/homebrew/Cellar/librsvg/2.50.3/lib -lrsvg-2 -lm -lgio-2.0 -lgdk_pixbuf-2.0 -lgobject-2.0 -lglib-2.0 -lintl -lcairo -L/opt/homebrew/Cellar/imagemagick/7.0.11-5/lib -lMagickWand-7.Q16HDRI -lMagickCore-7.Q16HDRI            -lxml2    -lncurses         -L/opt/homebrew/Cellar/gnutls/3.6.15/lib -lgnutls   -L/opt/homebrew/Cellar/little-cms2/2.12/lib -llcms2   -lz   -L/opt/homebrew/Cellar/jansson/2.13.1/lib -ljansson -lgmp
../lib-src/make-fingerprint temacs.tmp
mv temacs.tmp temacs
.././build-aux/install-sh -c -d ../etc
/Library/Developer/CommandLineTools/usr/bin/make -C ../lisp update-subdirs
for file in . ./calc ./calendar ./emacs-lisp ./emulation ./erc ./eshell ./gnus ./image ./international ./language ./mail ./mh-e ./net ./nxml ./obsolete ./org ./play ./progmodes ./term ./textmodes ./url ./vc; do \
           ./../build-aux/update-subdirs $file; \
        done;
cp -f temacs bootstrap-emacs
rm -f bootstrap-emacs.pdmp
./temacs --batch  -l loadup --temacs=pbootstrap
make[1]: *** [bootstrap-emacs.pdmp] Killed: 9
make: *** [src] Error 2

Do not report this issue to Homebrew/brew or Homebrew/core!
  </pre>
</details>

## Output of `brew config`

<pre>
$ arch -arm64 brew config
HOMEBREW_VERSION: 3.0.10
ORIGIN: https://github.com/Homebrew/brew
HEAD: 0ad2564c5f1ddf6f16f707f2861556b80546713c
Last commit: 5 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 72760b921df2bdd27a64fe0db5259f9e39a05338
Core tap last commit: 13 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.3 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit arm_firestorm_icestorm
Clang: 12.0 build 1200
Git: 2.31.1 => /opt/homebrew/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 11.2.2-arm64
CLT: 12.4.0.0.1.1610135815
Xcode: N/A
Rosetta 2: false
</pre>


## Output of `brew doctor`

<pre>
$ arch -arm64 brew doctor
Your system is ready to brew
</pre>

--- a/Formula/emacs-plus@27.rb
+++ b/Formula/emacs-plus@27.rb
@@ -103,6 +103,15 @@ class EmacsPlusAT27 < EmacsBase
     sha256 "782a222505ceea31f9032ed55e24dcbd0357b1178b916b536d3eb222c9dc1225"
   end
 
+  stable do
+    # Back-ported patch for configure and configure.guess to allow configure to complete
+    # for aarch64-apple-darwin targets.
+    patch do
+      url (UrlResolver.patch_url "emacs-27/arm")
+      sha256 "eeed4c0de763c08810d184501b64de3653adb84cc651497bdb4158ba3448312a"
+    end
+  end
+
   def install
     args = %W[
       --disable-dependency-tracking

--- a/patches/emacs-27/arm.patch
+++ b/patches/emacs-27/arm.patch
@@ -1,0 +1,165 @@
+diff -ur a/build-aux/config.guess b/build-aux/config.guess
+--- a/build-aux/config.guess	2020-07-29 15:40:41.000000000 -0600
++++ b/build-aux/config.guess	2020-11-25 15:44:16.000000000 -0700
+@@ -99,6 +99,8 @@
+ trap 'test -z "$tmp" || rm -fr "$tmp"' 0 1 2 13 15
+
+ set_cc_for_build() {
++    # prevent multiple calls if $tmp is already set
++    test "$tmp" && return 0
+     : "${TMPDIR=/tmp}"
+     # shellcheck disable=SC2039
+     { tmp=`(umask 077 && mktemp -d "$TMPDIR/cgXXXXXX") 2>/dev/null` && test -n "$tmp" && test -d "$tmp" ; } ||
+@@ -402,7 +404,7 @@
+ 	# If there is a compiler, see if it is configured for 64-bit objects.
+ 	# Note that the Sun cc does not turn __LP64__ into 1 like gcc does.
+ 	# This test works for both compilers.
+-	if [ "$CC_FOR_BUILD" != no_compiler_found ]; then
++	if test "$CC_FOR_BUILD" != no_compiler_found; then
+ 	    if (echo '#ifdef __amd64'; echo IS_64BIT_ARCH; echo '#endif') | \
+ 		(CCOPTS="" $CC_FOR_BUILD -E - 2>/dev/null) | \
+ 		grep IS_64BIT_ARCH >/dev/null
+@@ -542,10 +544,10 @@
+     AViiON:dgux:*:*)
+ 	# DG/UX returns AViiON for all architectures
+ 	UNAME_PROCESSOR=`/usr/bin/uname -p`
+-	if [ "$UNAME_PROCESSOR" = mc88100 ] || [ "$UNAME_PROCESSOR" = mc88110 ]
++	if test "$UNAME_PROCESSOR" = mc88100 || test "$UNAME_PROCESSOR" = mc88110
+ 	then
+-	    if [ "$TARGET_BINARY_INTERFACE"x = m88kdguxelfx ] || \
+-	       [ "$TARGET_BINARY_INTERFACE"x = x ]
++	    if test "$TARGET_BINARY_INTERFACE"x = m88kdguxelfx || \
++	       test "$TARGET_BINARY_INTERFACE"x = x
+ 	    then
+ 		echo m88k-dg-dgux"$UNAME_RELEASE"
+ 	    else
+@@ -578,7 +580,7 @@
+ 	echo i386-ibm-aix
+ 	exit ;;
+     ia64:AIX:*:*)
+-	if [ -x /usr/bin/oslevel ] ; then
++	if test -x /usr/bin/oslevel ; then
+ 		IBM_REV=`/usr/bin/oslevel`
+ 	else
+ 		IBM_REV="$UNAME_VERSION.$UNAME_RELEASE"
+@@ -618,7 +620,7 @@
+ 	else
+ 		IBM_ARCH=powerpc
+ 	fi
+-	if [ -x /usr/bin/lslpp ] ; then
++	if test -x /usr/bin/lslpp ; then
+ 		IBM_REV=`/usr/bin/lslpp -Lqc bos.rte.libc |
+ 			   awk -F: '{ print $3 }' | sed s/[0-9]*$/0/`
+ 	else
+@@ -653,7 +655,7 @@
+ 	    9000/31?)            HP_ARCH=m68000 ;;
+ 	    9000/[34]??)         HP_ARCH=m68k ;;
+ 	    9000/[678][0-9][0-9])
+-		if [ -x /usr/bin/getconf ]; then
++		if test -x /usr/bin/getconf; then
+ 		    sc_cpu_version=`/usr/bin/getconf SC_CPU_VERSION 2>/dev/null`
+ 		    sc_kernel_bits=`/usr/bin/getconf SC_KERNEL_BITS 2>/dev/null`
+ 		    case "$sc_cpu_version" in
+@@ -667,7 +669,7 @@
+ 			esac ;;
+ 		    esac
+ 		fi
+-		if [ "$HP_ARCH" = "" ]; then
++		if test "$HP_ARCH" = ""; then
+ 		    set_cc_for_build
+ 		    sed 's/^		//' << EOF > "$dummy.c"
+
+@@ -706,7 +708,7 @@
+ 		    test -z "$HP_ARCH" && HP_ARCH=hppa
+ 		fi ;;
+ 	esac
+-	if [ "$HP_ARCH" = hppa2.0w ]
++	if test "$HP_ARCH" = hppa2.0w
+ 	then
+ 	    set_cc_for_build
+
+@@ -780,7 +782,7 @@
+ 	echo hppa1.0-hp-osf
+ 	exit ;;
+     i*86:OSF1:*:*)
+-	if [ -x /usr/sbin/sysversion ] ; then
++	if test -x /usr/sbin/sysversion ; then
+ 	    echo "$UNAME_MACHINE"-unknown-osf1mk
+ 	else
+ 	    echo "$UNAME_MACHINE"-unknown-osf1
+@@ -924,7 +926,7 @@
+ 	echo "$UNAME_MACHINE"-unknown-linux-"$LIBC"
+ 	exit ;;
+     alpha:Linux:*:*)
+-	case `sed -n '/^cpu model/s/^.*: \(.*\)/\1/p' < /proc/cpuinfo` in
++	case `sed -n '/^cpu model/s/^.*: \(.*\)/\1/p' /proc/cpuinfo 2>/dev/null` in
+ 	  EV5)   UNAME_MACHINE=alphaev5 ;;
+ 	  EV56)  UNAME_MACHINE=alphaev56 ;;
+ 	  PCA56) UNAME_MACHINE=alphapca56 ;;
+@@ -1093,7 +1095,17 @@
+ 	echo "$UNAME_MACHINE"-dec-linux-"$LIBC"
+ 	exit ;;
+     x86_64:Linux:*:*)
+-	echo "$UNAME_MACHINE"-pc-linux-"$LIBC"
++	set_cc_for_build
++	LIBCABI=$LIBC
++	if test "$CC_FOR_BUILD" != no_compiler_found; then
++	    if (echo '#ifdef __ILP32__'; echo IS_X32; echo '#endif') | \
++		(CCOPTS="" $CC_FOR_BUILD -E - 2>/dev/null) | \
++		grep IS_X32 >/dev/null
++	    then
++		LIBCABI="$LIBC"x32
++	    fi
++	fi
++	echo "$UNAME_MACHINE"-pc-linux-"$LIBCABI"
+ 	exit ;;
+     xtensa*:Linux:*:*)
+ 	echo "$UNAME_MACHINE"-unknown-linux-"$LIBC"
+@@ -1282,7 +1294,7 @@
+ 	echo mips-sony-newsos6
+ 	exit ;;
+     R[34]000:*System_V*:*:* | R4000:UNIX_SYSV:*:* | R*000:UNIX_SV:*:*)
+-	if [ -d /usr/nec ]; then
++	if test -d /usr/nec; then
+ 		echo mips-nec-sysv"$UNAME_RELEASE"
+ 	else
+ 		echo mips-unknown-sysv"$UNAME_RELEASE"
+@@ -1330,6 +1342,9 @@
+     *:Rhapsody:*:*)
+ 	echo "$UNAME_MACHINE"-apple-rhapsody"$UNAME_RELEASE"
+ 	exit ;;
++    arm64:Darwin:*:*)
++	echo aarch64-apple-darwin"$UNAME_RELEASE"
++	exit ;;
+     *:Darwin:*:*)
+ 	UNAME_PROCESSOR=`uname -p`
+ 	case $UNAME_PROCESSOR in
+@@ -1344,7 +1359,7 @@
+ 	else
+ 	    set_cc_for_build
+ 	fi
+-	if [ "$CC_FOR_BUILD" != no_compiler_found ]; then
++	if test "$CC_FOR_BUILD" != no_compiler_found; then
+ 	    if (echo '#ifdef __LP64__'; echo IS_64BIT_ARCH; echo '#endif') | \
+ 		   (CCOPTS="" $CC_FOR_BUILD -E - 2>/dev/null) | \
+ 		   grep IS_64BIT_ARCH >/dev/null
+@@ -1627,6 +1642,12 @@
+   https://git.savannah.gnu.org/gitweb/?p=config.git;a=blob_plain;f=config.guess
+ and
+   https://git.savannah.gnu.org/gitweb/?p=config.git;a=blob_plain;f=config.sub
++EOF
++
++year=`echo $timestamp | sed 's,-.*,,'`
++# shellcheck disable=SC2003
++if test "`expr "\`date +%Y\`" - "$year"`" -lt 3 ; then
++   cat >&2 <<EOF
+
+ If $0 has already been updated, send the following data and any
+ information you think might be pertinent to config-patches@gnu.org to
+@@ -1654,5 +1675,6 @@
+ UNAME_SYSTEM  = "$UNAME_SYSTEM"
+ UNAME_VERSION = "$UNAME_VERSION"
+ EOF
++fi
+
+ exit 1

